enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,471 @@
+what the intent/reason for the hook is.
+Each hook name should be unique and managed by a single entity.
+
+##### owner (Optional)
+Some information about who created or is otherwise in charge of managing the annotation.
+This might be a controller or a username to indicate an administrator applied the hook directly.
+
+##### Annotation Examples
+
+These examples are all hypothetical to illustrate what form annotations should
+take. The names of each hook and the respective controllers are fictional.
+
+pre-drain.hook.machine.cluster-api.x-k8s.io/migrate-important-app: my-app-migration-controller
+
+pre-terminate.hook.machine.cluster-api.x-k8s.io/backup-files: my-backup-controller
+
+pre-terminate.hook.machine.cluster-api.x-k8s.io/wait-for-storage-detach: my-custom-storage-detach-controller
+
+#### Changes to machine-controller
+
+The machine controller should check for the existence of 1 or more hooks at
+specific points (lifecycle-points) during reconciliation.  If a hook matching
+the lifecycle-point is discovered, the machine-controller should stop
+reconciling the machine.
+
+##### Reconciliation
+
+When a Hook Implementing Controller updates the Machine, reconciliation will be
+triggered, and the Machine will continue reconciling as normal, unless another
+hook is still present; there is no need to 'fail' the reconciliation to
+enforce requeuing.
+
+When all hooks for a given lifecycle-point are removed, reconciliation
+will continue as normal.
+
+##### Preventing additions after Machine deletion
+
+Adding new hooks once the deletion process has been initiated may not behave as
+expected, for example if a pre-drain hook is added after the Machine has already
+been terminated.
+
+To prevent potentially confusing behaviour, once a Machine has been marked for deletion,
+it is expected that hook implementing controllers should not attempt to add new deletion
+hooks. The hook implementing controller is expected to detect the deletion in progress
+and perform any clean up logic necessary.
+
+To prevent potential issues here, and to ensure users do not accidentally add hooks,
+we will leverage the existing Machine API webhooks to prevent new annotations of the
+deletion hook format from being added to Machines after they have been marked for
+deletion.
+
+##### Hook failure
+
+The machine controller should not timeout or otherwise consider the lifecycle
+hook as 'failed.'  Only the Hook Implementing Controller (or the end user in
+extenuating circumstances) may decide to remove a particular lifecycle hook
+to allow the machine controller to progress past the corresponding lifecycle-point.
+
+The Machine status will contain conditions identifying why the Machine controller
+has not progressed in removing the Machine. It is expected that hook implementing
+controllers should signal to users when they are having issues via some mechanism
+(eg. their `ClusterOperator` status) and that the conditions on the Machine should
```

Comment:
The clusteroperator status isn't likely to be an entry point given the cardinality differences between the hook provider and the number of machines.

Is there any mechanism to be included in the machine status.  These hooks are pretty deep and condition names can be scope limited.  What about encouraging these hooks to place a condition aligned to their name when they fail?